File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:

        if not head:
            return None
        slow, fast = head,head
        for i in range(n):
            fast = fast.next
            if not fast: # only one element
                return head.next
        while fast.next:
            slow = slow.next
            fast = fast.next
        slow.next = slow.next.next
        return head
            
'''
    1,2,3,4,5
    1 -> remove 5

    first = 2


'''            

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int sz=0;
        ListNode dummy(0);
        dummy.next = head;
        ListNode * tmp = head;
        while(tmp) {
            tmp = tmp->next;
            sz++;
        }
        tmp = &dummy;
        sz -= n;
        while(sz) {
            tmp = tmp->next;
            sz--;
        }
        if(tmp->next)
        {
            tmp->next = tmp->next->next;
        }
        return dummy.next;
    }
};


/*
brute force

jump to nth - 1 node

nth-1 node 0=-> next->next = nth -> next

*/